src/algorithms/pagerank.c

```diff
@@ -0,0 +1,224 @@
+
+	//--------------------------------------------------------------------------
+	// iterate to compute the pagerank of each node
+	//--------------------------------------------------------------------------
+
+	for((*iters) = 0 ; (*iters) < itermax && rdiff > ftol ; (*iters)++) {
+
+		//----------------------------------------------------------------------
+		// t = (r*C or C*r) + (teleport * sum (r)) ;
+		//----------------------------------------------------------------------
+		// GxB_print (r, 2) ;
+
+		assert(GrB_reduce(&rsum, NULL, GxB_PLUS_FP32_MONOID, r, NULL) == GrB_SUCCESS) ;
+
+		// t = C*r
+		// using the transpose of A, scaled (dot product)
+		assert(GrB_mxv(t, NULL, NULL, GxB_PLUS_TIMES_FP32, C, r, NULL) == GrB_SUCCESS) ;
+
+		// t += teleport_scalar ;
+		float teleport_scalar = teleport * rsum ;
+		assert(GrB_assign(t, NULL, GrB_PLUS_FP32, teleport_scalar, GrB_ALL, n, NULL) == GrB_SUCCESS) ;
+		//----------------------------------------------------------------------
+		// rdiff = sum ((r-t).^2)
+		//----------------------------------------------------------------------
+
+		assert(GrB_eWiseAdd(r, NULL, NULL, op_diff, r, t, NULL) == GrB_SUCCESS) ;
+		assert(GrB_reduce(&rdiff, NULL, GxB_PLUS_FP32_MONOID, r, NULL) == GrB_SUCCESS) ;
+
+		//----------------------------------------------------------------------
+		// swap r and t
+		//----------------------------------------------------------------------
+
+		GrB_Vector temp = r ;
+		r = t ;
+		t = temp ;
+	}
+
+	assert(GrB_free(&C) == GrB_SUCCESS) ;
+	assert(GrB_free(&t) == GrB_SUCCESS) ;
+
+	//--------------------------------------------------------------------------
+	// scale the result
+	//--------------------------------------------------------------------------
+
+	// rsum = sum (r)
+	assert(GrB_reduce(&rsum, NULL, GxB_PLUS_FP32_MONOID, r, NULL) == GrB_SUCCESS) ;
+
+	// r = r / rsum
+	assert(GrB_Vector_assign_FP32(r, NULL, GrB_TIMES_FP32, 1 / rsum, GrB_ALL, n, NULL) == GrB_SUCCESS) ;
+
+	//--------------------------------------------------------------------------
+	// sort the nodes by pagerank
+	//--------------------------------------------------------------------------
+
+	// [r,irank] = sort (r, 'descend') ;
+
+	// extract the contents of r
+	assert(GxB_Vector_export(&r, &type, &n, &nvals, &I, (void **)(&X), NULL) == GrB_SUCCESS) ;
+
+	// this will always be true since r is dense, but check anyway:
+	if(nvals != n) return (GrB_PANIC) ;
+
+	// P = struct (X,I)
+	P = rm_malloc(n * sizeof(LAGraph_PageRank)) ;
+	assert(P != NULL);
+
+	for(int64_t k = 0 ; k < nvals ; k++) {
+		// The kth ranked page is P[k].page (with k=0 being the highest rank),
+		// and its pagerank is P[k].pagerank.
+		P [k].pagerank = X [k] ;
+		// I [k] == k will be true for SuiteSparse:GraphBLAS but in general I
+		// can be returned in any order, so use I [k] instead of k, for other
+		// GraphBLAS implementations.
+		P [k].page = k ;
+	}
+
+	// free workspace
```

Comment:
Delete comment.

Additionally, I definitely don't understand everything in this file yet, but I read the diff of the LAGraph implementation and all of the changes seem sensible to me!

Reply:
👍